src/claimtriedb.cpp

```diff
@@ -0,0 +1,215 @@
+template <typename K, typename V>
+struct CActualBuffer : public CAbstractBuffer
+{
+    CActualBuffer() : data() {}
+    typedef std::list<std::pair<K, V> > dataType;
+    dataType data;
+    typename dataType::iterator find(const K& key)
+    {
+        typename dataType::iterator itData = data.begin();
+        while (itData != data.end()) {
+            if (key == itData->first) break;
+            ++itData;
+        }
+        return itData;
+    }
+    typename dataType::iterator end()
+    {
+        return data.end();
+    }
+    typename dataType::iterator push_back(const std::pair<K, V>& element)
+    {
+        return data.insert(data.end(), element);
+    }
+    void write(const unsigned char key, CClaimTrieDb* db)
+    {
+        for (typename dataType::iterator it = data.begin(); it != data.end(); ++it) {
+            if (it->second.empty()) {
+                db->Erase(std::make_pair(key, it->first));
+            } else {
+                db->Write(std::make_pair(key, it->first), it->second);
+            }
+        }
+        data.clear();
+    }
+};
+
+CClaimTrieDb::CClaimTrieDb(bool fMemory, bool fWipe)
+                    : CDBWrapper(GetDataDir() / "claimtrie", 100, fMemory, fWipe, false)
+{
+}
+
+CClaimTrieDb::~CClaimTrieDb()
+{
+    for (std::map<unsigned char, CAbstractBuffer*>::iterator itQueue = queues.begin(); itQueue != queues.end(); ++itQueue) {
+        delete itQueue->second;
+    }
+}
+
+void CClaimTrieDb::writeQueues()
+{
+    for (std::map<unsigned char, CAbstractBuffer*>::iterator itQueue = queues.begin(); itQueue != queues.end(); ++itQueue) {
+        itQueue->second->write(itQueue->first, this);
+    }
+}
+
+template <typename K, typename V>
+bool CClaimTrieDb::getQueueRow(const K& key, V& row) const
+{
+    const unsigned char hash = hashType<K, V>();
+    std::map<unsigned char, CAbstractBuffer*>::const_iterator itQueue = queues.find(hash);
+    if (itQueue != queues.end()) {
+        CActualBuffer<K, V>* map = static_cast<CActualBuffer<K, V>*>(itQueue->second);
+        typename CActualBuffer<K, V>::dataType::const_iterator itData = map->find(key);
+        if (itData != map->end()) {
+            row = itData->second;
+            return true;
+        }
+    }
+    return Read(std::make_pair(hash, key), row);
+}
+
+template <typename K, typename V>
+void CClaimTrieDb::updateQueueRow(const K& key, V& row)
+{
+    const unsigned char hash = hashType<K, V>();
+    std::map<unsigned char, CAbstractBuffer*>::iterator itQueue = queues.find(hash);
+    if (itQueue == queues.end()) {
+        itQueue = queues.insert(itQueue, std::pair<unsigned char, CAbstractBuffer*>(hash, new CActualBuffer<K, V>));
+    }
+    CActualBuffer<K, V>* map = static_cast<CActualBuffer<K, V>*>(itQueue->second);
+    typename CActualBuffer<K, V>::dataType::iterator itData = map->find(key);
+    if (itData == map->end()) {
+        itData = map->push_back(std::make_pair(key, V()));
+    }
+    std::swap(itData->second, row);
+}
```

Comment:
If you were using a real map here, the above four lines would turn into a single `map->at(key) = row;`. That seems more straightforward to me. Are you intentionally trying to return the old row here? If so, it should use the method return. The way that's coded is going to lead to unintended consequences.

Reply:
Yeah, but i don't think it's worst, no?

Reply:
I do think it's pretty bad; it's what caused the problem that Kay found, right?

Reply:
I don't have good reason for swap, but follows current behavior.